### PR TITLE
ENH: truncate output of Groupby.groups

### LINKED ISSUE #1135: df.groupby('key').groups printed all: problem with large arrays
When you are working with a large array, it is not printed out in its entirety in the console, but when you have grouped them with `groupby('key')`, the groups are all printed out. Would it be possible to also restrict the output of `groupby('key').groups` to eg the first and last groups?

I was working with a rather large dataframe (around 80000 rows), and first it took a long time to print it all, and second the console got stuck for a while (but that could also be an issue with spyder).
I know it is not very useful to print out the groups, but I was curious to see what it would look like, but it was not a very good idea with such a large array.

```python
>>> import numpy as np
>>> import pandas as pd
>>> df = pd.DataFrame(np.random.randn(100000, 4), columns=list('abcd'))
>>> df['g'] = np.random.randint(0, 100, 100000)
>>> df.groupby('g').groups  # this repr should be truncated
```

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Resurrection of #24853. Have resolved conflicts and updated according to the last review it received. The original PR fixed some unrelated minor typos, so I've kept them in (and added a fix for another absolutely minor typo I'd noticed)

- [x] closes #1135
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4790,7 +4791,7 @@ def groupby(self, values) -> Dict[Hashable, np.ndarray]:
         # map to the label
         result = {k: self.take(v) for k, v in result.items()}
 
-        return result
+        return PrettyDict(result)
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py]: could also update return type on L4769 and docstring. would need PrettyDict to be generic, see other comment.

File: pandas/io/formats/printing.py [hunk 2]
```diff
@@ -528,3 +528,10 @@ def format_object_attrs(
     if len(obj) > max_seq_items:
         attrs.append(("length", len(obj)))
     return attrs
+
+
+class PrettyDict(dict):
+    """Dict extension to support abbreviated __repr__"""
+
+    def __repr__(self):
```
Comment (simonjayhawkins) [inline pandas/io/formats/printing.py]:
```suggestion
    def __repr__(self) -> str:
```

File: pandas/io/formats/printing.py [hunk 3]
```diff
@@ -528,3 +528,10 @@ def format_object_attrs(
     if len(obj) > max_seq_items:
         attrs.append(("length", len(obj)))
     return attrs
+
+
+class PrettyDict(dict):
```
Comment (simonjayhawkins) [inline pandas/io/formats/printing.py]: with mypy -disallow-any-generic, see #30539, this gives `pandas\io\formats\printing.py:533: error: Implicit generic "Any". Use "typing.Dict" and specify generic parameters`
```suggestion
class PrettyDict(Dict[_KT, _VT]):
```

and add
```python
_KT = TypeVar("_KT")
_VT = TypeVar("_VT")
```

File: pandas/tests/groupby/test_groupby.py [hunk 4]
```diff
@@ -2037,3 +2037,16 @@ def test_groupby_list_level():
...
     result = expected.groupby(level=[0]).mean()
     tm.assert_frame_equal(result, expected)
+
+
+def test_groups_repr_truncates():
+    # GH 1135
+    df = pd.DataFrame({"a": [1, 1, 1, 2, 2, 3], "b": [1, 2, 3, 4, 5, 6]})
+
+    with pd.option_context("display.max_rows", 2):
+        x = df.groupby("a").groups
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_groupby.py]: This is giving `{1: [0, 1, ...], 2: [3, 4], ...}` so altering max_seq_items is not ideal. Is there another way?

File: pandas/tests/groupby/test_groupby.py [hunk 5]
```diff
@@ -2037,3 +2037,16 @@ def test_groupby_list_level():
...
+        assert x.__repr__().endswith("...}")
```
Comment (simonjayhawkins) [inline pandas/tests/groupby/test_groupby.py]: can the assertion be more explicit instead of .endswith?

**Review by simonjayhawkins — COMMENTED** [review]
@MarcoGorelli a few comments, the typing ones are not blockers.

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/tests/groupby/test_groupby.py]: @simonjayhawkins could test this on a longer DataFrame and then not alter `display.max_rows`, would that be OK?

Comment (simonjayhawkins) [reply pandas/tests/groupby/test_groupby.py]: use `with pd.option_context("display.max_seq_items", 2):` and just `return pprint_thing(self)`  from `repr` instead?

File: pandas/core/indexes/base.py [hunk 6]
```diff
@@ -4765,7 +4766,7 @@ def _maybe_promote(self, other):
                 return self.astype("object"), other.astype("object")
         return self, other
 
-    def groupby(self, values) -> Dict[Hashable, np.ndarray]:
+    def groupby(self, values) -> PrettyDict:
```
Comment (simonjayhawkins) [inline pandas/core/indexes/base.py]: now PrettyDict is generic can do
```suggestion
    def groupby(self, values) -> PrettyDict[Hashable, np.ndarray]:
```

File: pandas/tests/groupby/test_groupby.py (hunk 4 above)
Comment (MarcoGorelli) [reply pandas/tests/groupby/test_groupby.py]: Thanks for your review - I've reread this sentence a few times

>  just return pprint_thing(self) from repr instead

but I don't understand what you mean. If I change the definition of `PrettyDict.__repr__` to
```
        return pprint_thing(self)
```
then won't the issue be not-solved? At least, with the test as it's currently written, it fails

Comment (simonjayhawkins) [reply pandas/tests/groupby/test_groupby.py]: I thought pprint_thing called_pprint_dict which uses display.max_seq_items as the default. I'll check in a bit.

Comment (simonjayhawkins) [reply pandas/tests/groupby/test_groupby.py]: > I thought pprint_thing called_pprint_dict which uses display.max_seq_items as the default. I'll check in a bit.

confirmed.

> then won't the issue be not-solved?

if the top level of the dict **must** follow display.max_rows for truncation, then maybe updating pprint_thing to accept, say, a tuple for max_seq_items to allow the nested sequences to follow display.max_seq_items could be an option. The tuple could be any length that specifies the max_seq_items for each level with the last value being used for deeper sequences.

so something like `return pprint_thing(self, max_seq_items=(get_option("display.max_rows"), get_option("display.max_seq_items"))` ?

or maybe alter the \_\_repr__ to be less dict-like

Comment (MarcoGorelli) [reply pandas/tests/groupby/test_groupby.py]: > At least, with the test as it's currently written, it fails

sorry, my mistake here, hadn't realised the default for max_seq_items is longer than that for max_rows, I should've checked before commenting. Have updated according to your original request

**Review by simonjayhawkins — APPROVED** [review]
Thanks @MarcoGorelli lgtm.

**Comment by jreback** [comment]
thanks @MarcoGorelli

**Pull request merged**